src/pipewire_listener.rs

```diff
@@ -52,6 +53,165 @@ enum PipewireObjectType {
     Link,
 }
 
+#[allow(dead_code)]
+#[derive(Debug, Clone)]
+pub enum PortConnections {
+    AllInOrder,
+    Only(Vec<(String, String)>),
+}
+
+#[derive(Debug, Clone)]
+pub struct StreamConnections {
+    pub output_stream: String,
+    pub input_stream: String,
+    pub port_connections: PortConnections,
+}
+
+#[derive(Debug)]
+pub struct StreamDescriptor {
+    pub name: String,
+    pub inputs: Vec<String>,
+    pub ouputs: Vec<String>,
+}
+
+pub struct PipewireController {
+    sender: pipewire::channel::Sender<Vec<StreamConnections>>,
+    state: Arc<Mutex<PipewireState>>,
+}
+
+impl PipewireController {
+    pub fn new() -> Self {
+        let (sender, receiver) = pipewire::channel::channel();
+        let state = Arc::new(Mutex::new(PipewireState::new()));
+        thread::spawn({
+            let state = state.clone();
+            move || pipewire_thread(vec![], receiver, Arc::clone(&state))
+        });
+        PipewireController {
+            sender,
+            state: Arc::clone(&state),
+        }
+    }
+
+    pub fn set_stream_connections(&self, stream_connections: Vec<StreamConnections>) -> Result<()> {
+        let result = self.sender.send(stream_connections);
+        if result.is_err() {
+            return Err(anyhow!("Failed to change pipewire connections"));
+        }
+        Ok(())
+    }
+
+    pub fn get_streams(&self) -> Result<Vec<StreamDescriptor>> {
+        let mut stream_descriptors = Vec::new();
+        let state = self.state.lock().unwrap();
+        for node in state.nodes.values() {
+            let mut descriptor = StreamDescriptor {
+                name: node.name.clone(),
+                inputs: vec![],
+                ouputs: vec![],
+            };
+            for input_id in &node.input_ports {
```

Comment:
Can you zip over the 2? Something like `for (input_id, output_id) in node.input_ports.iter().zip(node.output_ports) { ... } `

Reply:
pourquoi

Reply:
on a pas le meme nombre de input ports et output ports